org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/Launcher.java

```diff
@@ -310,28 +312,61 @@
 			} else {
 				remoteProxy = (T) ServiceEndpoints.toServiceObject(remoteEndpoint, (Collection<Class<?>>) (Object) remoteInterfaces, classLoader);
 			}
+			
+			// create the message processor
 			StreamMessageProducer reader = new StreamMessageProducer(input, jsonHandler, remoteEndpoint);
 			MessageConsumer messageConsumer = wrapMessageConsumer(remoteEndpoint);
 			ExecutorService execService = executorService != null ? executorService : Executors.newCachedThreadPool();
+			ConcurrentMessageProcessor<T> msgProcessor = createMessageProcessor(reader, messageConsumer, remoteProxy);
 			
-			return new Launcher<T> () {
-				@Override
-				public Future<Void> startListening() {
-					return ConcurrentMessageProcessor.startProcessing(reader, messageConsumer, execService);
-				}
+			return createLauncher(reader, messageConsumer, execService, remoteProxy, remoteEndpoint, msgProcessor);
+		}
+		
+		protected Launcher<T> createLauncher(StreamMessageProducer reader, MessageConsumer messageConsumer,
+				ExecutorService execService, T remoteProxy, RemoteEndpoint remoteEndpoint, ConcurrentMessageProcessor<T> msgProcessor) {
+			return new StandardLauncher<T>(reader, execService, remoteProxy, remoteEndpoint, msgProcessor);
+		}
+		
+		public static class StandardLauncher<T> implements Launcher<T> {
+			private StreamMessageProducer reader;
+			private ExecutorService execService;
+			private T remoteProxy;
+			private RemoteEndpoint remoteEndpoint;
+			private ConcurrentMessageProcessor<T> msgProcessor;
 
-				@Override
-				public T getRemoteProxy() {
-					return remoteProxy;
-				}
+			public StandardLauncher(StreamMessageProducer reader, MessageConsumer messageConsumer,
+					ExecutorService execService, T remoteProxy, RemoteEndpoint remoteEndpoint) {
+				this(reader, execService, remoteProxy, remoteEndpoint, 
+						new ConcurrentMessageProcessor<>(reader, messageConsumer));
+			}
+			
+			public StandardLauncher(StreamMessageProducer reader2,
+					ExecutorService execService2, T remoteProxy2, RemoteEndpoint remoteEndpoint2,
+					ConcurrentMessageProcessor<T> msgProcessor) {
+				this.reader = reader2;
+				this.execService = execService2;
+				this.remoteProxy = remoteProxy2;
+				this.remoteEndpoint = remoteEndpoint2;
+				this.msgProcessor = msgProcessor;
+			}
 
-				@Override
-				public RemoteEndpoint getRemoteEndpoint() {
-					return remoteEndpoint;
-				}
-			};
+			@Override
+			public Future<Void> startListening() {
+				final Future<?> result = execService.submit(msgProcessor);
+				return ConcurrentMessageProcessor.wrapFuture(result, reader);
```

Comment:
You could call the new `beginProcessing(…)` method here (otherwise we could as well remove it).